File: packages/generator-core/src/instance.ts
import { FormInputError } from '@crudui/validator';
import { bindForm, copyFormValue, type BindFormOptions, type FormFieldTemplate, type FormTemplate } from './form';
import { bindButtons, type ButtonVM } from './buttons';
import { formMessages, type FormMessages } from './messages';
import type { NodeVM } from './viewmodel';
import { getValueByPath, parsePathString } from './util';
import { canUndo, canRedo, emptyHistory, recordChange, undoChange, redoChange, type History } from './history';
import { initialView, rekeyRowView, removeRowView, setAllExpandedView, toggleRowView, type ViewState } from './view';
import { checkFormTemplate } from './template-shape';
import { checkArgumentText, checkBindText } from './input-text';

/** Transport key for an existing database sequence. */
export function sequenceRowKey(sequence: string | number | bigint): string {
  if (typeof sequence === 'number' && !Number.isSafeInteger(sequence)) {
    throw new FormInputError('A sequence must be a safe integer or a decimal string');
  }
  const digits = String(sequence);
  if (!/^\d{1,13}$/.test(digits)) throw new FormInputError('A sequence must contain 1–13 decimal digits');
  return `__${digits.padStart(13, '0')}__`;
}

/** Generate a new row key. This is identity, never a serialization index. */
export function createRowKey(): string {
  const bytes = globalThis.crypto.getRandomValues(new Uint8Array(7));
  return `__${Array.from(bytes, b => b.toString(16).padStart(2, '0')).join('').slice(0, 13)}__`;
}

/** Per-instance presentation options. */
export type CreateFormOptions = Pick<BindFormOptions, 'language' | 'keyPrefix' | 'idPrefix'>;

/** Values and insertion position for a new repeated row. */
export interface AddRowOptions {
  /** Explicit row key; omitted to generate one. */
  key?: string;
  /** Insert after this existing key; omitted to append. */
  afterKey?: string;
  /** Initial values; omitted to use field defaults. */
  value?: unknown;
}

/** Stable subscription snapshot, changed only after a successful operation or view change. */
export interface FormSnapshot {
  /** Evaluated fields for the current data and view. */
  readonly fields: NodeVM[];
  /** Evaluated form buttons for the current data. */
  readonly buttons: ButtonVM[];
  /** Number of successful data updates. */
  readonly revision: number;
  /** Whether `undo` can restore an earlier record. */
  readonly canUndo: boolean;
  /** Whether `redo` can reapply a reverted change. */
  readonly canRedo: boolean;
}

function hasOwn(value: object, key: string): boolean {
  return Object.prototype.hasOwnProperty.call(value, key);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value) &&
    (Object.getPrototypeOf(value) === Object.prototype || Object.getPrototypeOf(value) === null);
}

function repeats(field: FormFieldTemplate): boolean {
  return field.spec.multiple === true || dataOnly(field) || isRecord(field.spec.multiple);
}

/** A `multiple: only` collection: its rows come only from the data. */
function dataOnly(field: FormFieldTemplate): boolean {
  const multiple = field.spec.multiple;
  return multiple === 'only' || (isRecord(multiple) && multiple.only === true);
}

function checkedSegments(path: string): string[] {
  const segments = parsePathString(path);
  if (!segments.length || segments.some(s => ['__proto__', 'prototype', 'constructor'].includes(s))) {
    throw new FormInputError(`Invalid form path: ${path}`);
  }
  return segments;
}

function checkKey(key: string): void {
  // Dot/bracket characters are path syntax. Numeric object keys cannot retain
  // insertion order in JavaScript; normalize database sequences explicitly.
  if (!/^[A-Za-z0-9_-]+$/.test(key) || /^\d+$/.test(key) ||
      ['__proto__', 'prototype', 'constructor'].includes(key)) {
    throw new FormInputError(`Invalid row key: ${key}; use sequenceRowKey for numeric ids`);
  }
}

function putAt(data: Record<string, unknown>, path: string[], value: unknown): Record<string, unknown> {
  const [head, ...tail] = path;
  return { ...data, [head]: tail.length
    ? putAt(isRecord(data[head]) ? data[head] : {}, tail, value) : value };
}

/** Data and view changes applied together by one commit. */
interface Change {
  /** Path of a value edit; consecutive edits of one path share an undo entry. */
  path?: string;
  /** Replacement view state. */
  view?: ViewState;
  /** Replace the record: clear history and view state. */
  reset?: boolean;
}

/**
 * One editable form instance over a shared template. All row operations are
 * scoped to a collection path; no global string replacement touches siblings.
 * Data, undo history and view state (collapsed rows) are separate; view state is
 * never submitted.
 */
export class FormInstance {
  private data: Record<string, unknown>;
  private snapshot: FormSnapshot;
  private readonly listeners = new Set<() => void>();
  private readonly options: CreateFormOptions;
  private history: History<Record<string, unknown>> = emptyHistory();
  private view: ViewState = initialView();

  /** Shared immutable structure. */
  readonly template: FormTemplate;

  /** Create an instance with independent record data. */
  constructor(template: FormTemplate, data: Record<string, unknown> = {}, options: CreateFormOptions = {}) {
    checkBindText(template, data, options);
    checkFormTemplate(template);
    this.template = template;
    this.options = { ...options };
    this.data = this.normalizeFields(template.fields, data);
    this.snapshot = {
      fields: this.build(this.data, this.view.collapsed),
      buttons: bindButtons(template, this.data, this.options),
      revision: 0,
      canUndo: false,
      canRedo: false,
    };
  }

  /** Subscribe to injection, editing, row operations and view changes. Returns an unsubscribe function. */
  subscribe = (listener: () => void): (() => void) => {
    this.listeners.add(listener);
    return () => { this.listeners.delete(listener); };
  };

  /** Return the current subscription snapshot. */
  getSnapshot = (): FormSnapshot => this.snapshot;

  /** Effective root prefix for submitted input names. */
  get keyPrefix(): string | undefined { return this.options.keyPrefix ?? this.template.keyPrefix; }

  /** Interface text for the instance language. */
  get messages(): FormMessages { return formMessages(this.options.language ?? 'ko'); }

  /** Detached submission data, retaining keyed nested collections. */
  getData(): Record<string, unknown> { return copyFormValue(this.data); }

  /** Read a detached value relative to the data root. */
  getValue(path: string): unknown {
    checkArgumentText([['path', path]]);
    checkedSegments(path);
    return copyFormValue(getValueByPath(this.data, path));
  }

  /** Replace the record after an asynchronous load or save; history and view state restart. */
  setData(data: Record<string, unknown>): void {
    checkArgumentText([['data', data]]);
    this.commit(this.normalizeFields(this.template.fields, data), { reset: true });
  }

  /** Update one data path and reevaluate affected form presentation. */
  setValue(path: string, value: unknown): void {
    checkArgumentText([['path', path], ['value', value]]);
    const segments = checkedSegments(path);
    this.commit(this.normalizeFields(this.template.fields, putAt(this.data, segments, copyFormValue(value))),
      { path: segments.join('.') });
  }

  /** Insert a blank/defaulted row, or supplied row data, with one new identity. */
  addRow(path: string, options: AddRowOptions = {}): string {
    checkArgumentText([['path', path]], options, ['afterKey', 'key', 'value']);
    const { field, rows } = this.editableCollection(path);
    const settings = isRecord(field.spec.multiple) ? field.spec.multiple : {};
    if (typeof settings.max === 'number' && Object.keys(rows).length >= settings.max) {
      throw new FormInputError(`Maximum row count reached: ${path}`);
    }
    const key = options.key ?? this.freshKey(new Set(Object.keys(rows)));
    checkKey(key);
    if (hasOwn(rows, key)) throw new FormInputError(`Row key already exists: ${key}`);
    const entries = Object.entries(rows);
    const at = options.afterKey === undefined ? entries.length : entries.findIndex(([k]) => k === options.afterKey) + 1;
    if (options.afterKey !== undefined && at === 0) throw new FormInputError(`Unknown row: ${options.afterKey}`);
    entries.splice(at, 0, [key, this.normalizeRow(field, options.value, [...checkedSegments(path), key].join('.'))]);
    this.commit(putAt(this.data, checkedSegments(path), Object.fromEntries(entries)));
    return key;
  }

  /** Copy current values and give every descendant repeated row a fresh key. */
  copyRow(path: string, key: string, options: Omit<AddRowOptions, 'value'> = {}): string {
    checkArgumentText([['path', path], ['key', key]], options, ['afterKey', 'key']);
    const { field, rows } = this.editableCollection(path);
    if (!hasOwn(rows, key)) throw new FormInputError(`Unknown row: ${key}`);
    const value = this.copyRowValue(field, rows[key]);
    return this.addRow(path, { ...options, afterKey: options.afterKey ?? key, value });
  }

  /** Remove one row unless the minimum count would be violated. */
  removeRow(path: string, key: string): void {
    checkArgumentText([['path', path], ['key', key]]);
    const { field, rows } = this.editableCollection(path);
    if (!hasOwn(rows, key)) throw new FormInputError(`Unknown row: ${key}`);
    const settings = isRecord(field.spec.multiple) ? field.spec.multiple : {};
    if (typeof settings.min === 'number' && Object.keys(rows).length <= settings.min) {
      throw new FormInputError(`Minimum row count reached: ${path}`);
    }
    const segments = checkedSegments(path);
    this.commit(putAt(this.data, segments, Object.fromEntries(Object.entries(rows).filter(([k]) => k !== key))), {
      view: removeRowView(this.view, [...segments, key].join('.')),
    });
  }

  /** Change order without changing row keys, values or descendant identities. */
  moveRow(path: string, key: string, toIndex: number): void {
    checkArgumentText([['path', path], ['key', key]]);
    const { rows } = this.editableCollection(path);
    const entries = Object.entries(rows);
    const from = entries.findIndex(([k]) => k === key);
    if (from === -1) throw new FormInputError(`Unknown row: ${key}`);
    if (!Number.isInteger(toIndex) || toIndex < 0 || toIndex >= entries.length) {
      throw new FormInputError(`Invalid row position: ${toIndex}`);
    }
    if (from === toIndex) return;
    const [row] = entries.splice(from, 1);
    entries.splice(toIndex, 0, row);
    this.commit(putAt(this.data, checkedSegments(path), Object.fromEntries(entries)));
  }

  /** Apply a saved seq key to exactly one row. Descendant paths and view state follow. */
  rekeyRow(path: string, oldKey: string, newKey: string): void {
    checkArgumentText([['path', path], ['oldKey', oldKey], ['newKey', newKey]]);
    const { rows } = this.editableCollection(path);
    checkKey(newKey);
    if (!hasOwn(rows, oldKey)) throw new FormInputError(`Unknown row: ${oldKey}`);
    if (oldKey === newKey) return;
    if (hasOwn(rows, newKey)) throw new FormInputError(`Row key already exists: ${newKey}`);
    const segments = checkedSegments(path);
    this.commit(putAt(this.data, segments,
      Object.fromEntries(Object.entries(rows).map(([key, value]) => [key === oldKey ? newKey : key, value]))), {
      view: rekeyRowView(this.view, [...segments, oldKey].join('.'), [...segments, newKey].join('.')),
    });
  }

  /** Expand a collapsed row or collapse an expanded one. */
  toggleRow(path: string, key: string): void {
    this.refreshView(toggleRowView(this.view, this.rowPath(path, key)));
  }

  /** Expand or collapse every collapsible row. */
  setAllExpanded(expanded: boolean): void {
    this.refreshView(setAllExpandedView(this.snapshot.fields, expanded));
  }

  /** Restore the record before the last data change. */
  undo(): void {
    const { history, value: previous } = undoChange(this.history, this.data);
    const fields = this.build(previous, this.view.collapsed);
    this.history = history;
    this.data = previous;
    this.publish(fields, this.snapshot.revision + 1);
  }

  /** Reapply the most recently undone data change. */
  redo(): void {
    const { history, value: next } = redoChange(this.history, this.data);
    const fields = this.build(next, this.view.collapsed);
    this.history = history;
    this.data = next;
    this.publish(fields, this.snapshot.revision + 1);
  }

  private build(data: Record<string, unknown>, collapsed: ReadonlySet<string>): NodeVM[] {
    return bindForm(this.template, data, { ...this.options, collapsed });
  }

  private publish(fields: NodeVM[], revision: number): void {
    this.snapshot = {
      fields,
      buttons: bindButtons(this.template, this.data, this.options),
      revision,
      canUndo: canUndo(this.history),
      canRedo: canRedo(this.history),
    };
    for (const listener of this.listeners) listener();
  }

  private refreshView(view: ViewState): void {
    const fields = this.build(this.data, view.collapsed);
    this.view = view;
    this.publish(fields, this.snapshot.revision);
  }

  private commit(data: Record<string, unknown>, change: Change = {}): void {
    const view = change.reset ? initialView() : change.view ?? this.view;
    const fields = this.build(data, view.collapsed); // Failure leaves data, history and view unchanged.
    this.history = change.reset ? emptyHistory() : recordChange(this.history, this.data, change.path);
    this.view = view;
    this.data = data;
    this.publish(fields, this.snapshot.revision + 1);
  }

  /** Canonical row path after checking that the row exists. */
  private rowPath(path: string, key: string): string {
    const { rows } = this.collection(path);
    if (!hasOwn(rows, key)) throw new FormInputError(`Unknown row: ${key}`);
    return [...checkedSegments(path), key].join('.');
  }

  private freshKey(used: Set<string>): string {
    for (let attempt = 0; attempt < 100; attempt++) {
      const key = createRowKey();
      checkKey(key);
      if (!used.has(key)) return key;
    }
    throw new FormInputError('Unable to generate an unused row key');
  }

  /** Normalize record data; `path` is the full data path, empty at the root. */
  private normalizeFields(fields: readonly FormFieldTemplate[], value: unknown, path = ''): Record<string, unknown> {
    if (value !== undefined && !isRecord(value)) {
      throw new FormInputError(path ? `Group data must be an object: ${path}` : 'Form data must be an object');
    }
    const data = value === undefined ? {} : copyFormValue(value as Record<string, unknown>);
    for (const field of fields) {
      const raw = data[field.name];
      const fieldPath = path ? `${path}.${field.name}` : field.name;
      if (repeats(field)) {
        const rows: Record<string, unknown> = {};
        // Missing data creates one usable prototype, except in a data-only collection. Explicit {} means zero rows.
        if (raw !== undefined && !isRecord(raw)) throw new FormInputError(`Repeated data must be a keyed object: ${fieldPath}`);
        const entries = raw === undefined ? (dataOnly(field) ? [] : [[this.freshKey(new Set()), undefined] as const])
          : Object.entries(raw as Record<string, unknown>);
        for (const [key, row] of entries) {
          checkKey(key);
          if (hasOwn(rows, key)) throw new FormInputError(`Duplicate normalized row key: ${key}`);
          rows[key] = this.normalizeRow(field, row, `${fieldPath}.${key}`);
        }
        data[field.name] = rows;
      } else if (field.spec.type === 'group') {
        data[field.name] = this.normalizeFields(field.children, raw, fieldPath);
      } else if (raw === undefined && field.spec.default !== undefined) {
        data[field.name] = copyFormValue(field.spec.default);
      }
    }
    return data;
  }

  private normalizeRow(field: FormFieldTemplate, value: unknown, path: string): unknown {
    return field.spec.type === 'group' ? this.normalizeFields(field.children, value, path)
      : copyFormValue(value === undefined ? field.spec.default ?? '' : value);
  }

  private copyRowValue(field: FormFieldTemplate, value: unknown): unknown {
    if (field.spec.type !== 'group') return copyFormValue(value);
    const row = isRecord(value) ? copyFormValue(value) : {};
    return this.copyChildren(field.children, row);
  }

  private copyChildren(fields: readonly FormFieldTemplate[], value: Record<string, unknown>): Record<string, unknown> {
    const out = { ...value };
    for (const child of fields) {
      const raw = out[child.name];
      if (repeats(child) && isRecord(raw)) {
        const used = new Set(Object.keys(raw));
        out[child.name] = Object.fromEntries(Object.values(raw).map(row => {
          const key = this.freshKey(used);
          used.add(key);
          return [key, this.copyRowValue(child, row)];
        }));
      } else if (child.spec.type === 'group' && isRecord(raw)) {
        out[child.name] = this.copyChildren(child.children, raw);
      }
    }
    return out;
  }

  /** A collection whose rows the form may add, copy, remove, move or rekey. */
  private editableCollection(path: string): { field: FormFieldTemplate; rows: Record<string, unknown> } {
    const found = this.collection(path);
    if (dataOnly(found.field)) throw new FormInputError(`Rows of ${path} come only from data`);
    return found;
  }

  private collection(path: string): { field: FormFieldTemplate; rows: Record<string, unknown> } {
    const segments = checkedSegments(path);
    let fields = this.template.fields;
    let field: FormFieldTemplate | undefined;
    for (let i = 0; i < segments.length; i++) {
      field = fields.find(f => f.name === segments[i]);
      if (!field) throw new FormInputError(`Unknown collection: ${path}`);
      if (i === segments.length - 1) break;
      if (repeats(field)) i++;
      fields = field.children;
      field = undefined;
    }
    const rows = getValueByPath(this.data, path);
    if (!field || !repeats(field) || !isRecord(rows)) throw new FormInputError(`Not a keyed collection: ${path}`);
    return { field, rows };
  }
}

/** Create an independent mutable instance from a cacheable form template. */
export function createForm(template: FormTemplate, data?: Record<string, unknown>, options?: CreateFormOptions): FormInstance {
  return new FormInstance(template, data, options);
}

File: packages/generator-rust/src/instance.rs
use crate::template::repeats;
use crate::util::{segments, value_at};
use crate::{bind_form, BindOptions, FieldTemplate, FormError, FormResult, FormTemplate};
use serde::{Deserialize, Serialize};
use serde_json::{json, Map, Value};

/// Format a nonnegative sequence containing at most 13 decimal digits.
pub fn sequence_row_key(sequence: &str) -> FormResult<String> {
    if sequence.is_empty() || sequence.len() > 13 || !sequence.bytes().all(|c| c.is_ascii_digit()) {
        return Err(FormError::input(
            "A sequence must contain 1–13 decimal digits",
        ));
    }
    Ok(format!("__{sequence:0>13}__"))
}

/// Generate a random 13-character hexadecimal row key using the OS random source.
pub fn create_row_key() -> FormResult<String> {
    let mut bytes = [0u8; 7];
    getrandom::fill(&mut bytes)
        .map_err(|e| FormError::input(format!("Unable to generate a row key: {e}")))?;
    let digits = bytes.iter().map(|b| format!("{b:02x}")).collect::<String>();
    Ok(format!("__{}__", &digits[..13]))
}

fn check_key(key: &str) -> FormResult<()> {
    if key.is_empty()
        || !key
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
        || key.bytes().all(|b| b.is_ascii_digit())
        || ["__proto__", "prototype", "constructor"].contains(&key)
    {
        return Err(FormError::input(format!(
            "Invalid row key: {key}; use sequenceRowKey for numeric ids"
        )));
    }
    Ok(())
}

fn checked_segments(path: &str) -> FormResult<Vec<String>> {
    let parts = segments(path);
    if parts.is_empty()
        || parts
            .iter()
            .any(|s| ["__proto__", "prototype", "constructor"].contains(&s.as_str()))
    {
        return Err(FormError::input(format!("Invalid form path: {path}")));
    }
    Ok(parts)
}

fn put_at(data: &Value, path: &[String], value: Value) -> Value {
    let mut object = data.as_object().cloned().unwrap_or_default();
    let head = &path[0];
    let next = if path.len() == 1 {
        value
    } else {
        put_at(object.get(head).unwrap_or(&Value::Null), &path[1..], value)
    };
    object.insert(head.clone(), next);
    object.into()
}

fn fresh_key(used: &Map<String, Value>) -> FormResult<String> {
    for _ in 0..100 {
        let key = create_row_key()?;
        if !used.contains_key(&key) {
            return Ok(key);
        }
    }
    Err(FormError::input("Unable to generate an unused row key"))
}

fn normalize_row(field: &FieldTemplate, value: Option<&Value>) -> FormResult<Value> {
    if field.spec.get("type").is_some_and(|t| t == "group") {
        normalize_fields(&field.children, value)
    } else {
        Ok(value
            .or_else(|| field.spec.get("default"))
            .cloned()
            .unwrap_or(json!("")))
    }
}

fn normalize_fields(fields: &[FieldTemplate], value: Option<&Value>) -> FormResult<Value> {
    let mut data = match value {
        None => Map::new(),
        Some(Value::Object(map)) => map.clone(),
        _ => return Err(FormError::input("Group data must be an object")),
    };
    for field in fields {
        let raw = data.get(&field.name);
        if repeats(field) {
            let mut rows = Map::new();
            match raw {
                None => {
                    rows.insert(fresh_key(&rows)?, normalize_row(field, None)?);
                }
                Some(Value::Object(map)) => {
                    for (key, value) in map {
                        check_key(key)?;
                        rows.insert(key.clone(), normalize_row(field, Some(value))?);
                    }
                }
                _ => {
                    return Err(FormError::input(format!(
                        "Repeated data must be a keyed object: {}",
                        field.name
                    )))
                }
            }
            data.insert(field.name.clone(), rows.into());
        } else if field.spec.get("type").is_some_and(|t| t == "group") {
            let group = normalize_fields(&field.children, raw)?;
            data.insert(field.name.clone(), group);
        } else if raw.is_none() {
            if let Some(default) = field.spec.get("default") {
                data.insert(field.name.clone(), default.clone());
            }
        }
    }
    Ok(data.into())
}

fn copy_row_value(field: &FieldTemplate, value: &Value) -> FormResult<Value> {
    if field.spec.get("type").is_none_or(|t| t != "group") {
        return Ok(value.clone());
    }
    let mut row = value.as_object().cloned().unwrap_or_default();
    for child in &field.children {
        let Some(raw) = row.get(&child.name).filter(|v| v.is_object()) else {
            continue;
        };
        if repeats(child) {
            let mut used = raw.as_object().unwrap().clone();
            let mut copied = Map::new();
            for value in raw.as_object().unwrap().values() {
                let key = fresh_key(&used)?;
                used.insert(key.clone(), Value::Null);
                copied.insert(key, copy_row_value(child, value)?);
            }
            row.insert(child.name.clone(), copied.into());
        } else if child.spec.get("type").is_some_and(|t| t == "group") {
            let copied = copy_row_value(child, raw)?;
            row.insert(child.name.clone(), copied);
        }
    }
    Ok(row.into())
}

/// Values, identity and insertion position for an added or copied row.
#[derive(Debug, Default, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddRowOptions {
    /// Supplied row key; omitted to generate a new key.
    pub key: Option<String>,
    /// Existing row after which insertion occurs; omitted to append.
    pub after_key: Option<String>,
    /// Initial value; omitted to apply field defaults.
    #[serde(
        default,
        deserialize_with = "present_value",
        skip_serializing_if = "Option::is_none"
    )]
    pub value: Option<Value>,
}

fn present_value<'de, D: serde::Deserializer<'de>>(
    deserializer: D,
) -> Result<Option<Value>, D::Error> {
    Value::deserialize(deserializer).map(Some)
}

/// An independent editable form instance using a compiled structure.
#[derive(Clone)]
pub struct Form {
    template: FormTemplate,
    data: Value,
    fields: Vec<Value>,
    options: BindOptions,
    revision: u64,
}

impl Form {
    /// Create an instance, prepare default data and evaluate fields.
    pub fn new(template: FormTemplate, data: &Value, options: BindOptions) -> FormResult<Self> {
        let data = normalize_fields(&template.fields, Some(data))?;
        let fields = bind_form(&template, &data, &options)?;
        Ok(Self {
            template,
            data,
            fields,
            options,
            revision: 0,
        })
    }

    /// Return the shared structure without instance values.
    pub fn template(&self) -> &FormTemplate {
        &self.template
    }

    /// Return the current field models.
    pub fn fields(&self) -> &[Value] {
        &self.fields
    }

    /// Number of successful instance updates.
    pub fn revision(&self) -> u64 {
        self.revision
    }

    /// Return detached submission data.
    pub fn get_data(&self) -> Value {
        self.data.clone()
    }

    /// Replace record data and reevaluate the form atomically.
    pub fn set_data(&mut self, data: &Value) -> FormResult<()> {
        self.commit(normalize_fields(&self.template.fields, Some(data))?)
    }

    /// Return a detached value relative to the data root.
    pub fn get_value(&self, path: &str) -> FormResult<Value> {
        checked_segments(path)?;
        Ok(value_at(&self.data, path).cloned().unwrap_or(Value::Null))
    }

    /// Update one data path and reevaluate the form atomically.
    pub fn set_value(&mut self, path: &str, value: Value) -> FormResult<()> {
        let data = put_at(&self.data, &checked_segments(path)?, value);
        self.set_data(&data)
    }

    fn commit(&mut self, data: Value) -> FormResult<()> {
        let fields = bind_form(&self.template, &data, &self.options)?;
        self.data = data;
        self.fields = fields;
        self.revision += 1;
        Ok(())
    }

    fn collection(&self, path: &str) -> FormResult<(&FieldTemplate, &Map<String, Value>)> {
        let parts = checked_segments(path)?;
        let mut fields = &self.template.fields;
        let mut selected = None;
        let mut i = 0;
        while i < parts.len() {
            let field = fields
                .iter()
                .find(|f| f.name == parts[i])
                .ok_or_else(|| FormError::input(format!("Unknown collection: {path}")))?;
            if i + 1 == parts.len() {
                selected = Some(field);
                break;
            }
            if repeats(field) {
                i += 1;
            }
            fields = &field.children;
            i += 1;
        }
        let field = selected
            .filter(|f| repeats(f))
            .ok_or_else(|| FormError::input(format!("Not a keyed collection: {path}")))?;
        let rows = value_at(&self.data, path)
            .and_then(Value::as_object)
            .ok_or_else(|| FormError::input(format!("Not a keyed collection: {path}")))?;
        Ok((field, rows))
    }

    /// Insert a row with supplied data or field defaults.
    pub fn add_row(&mut self, path: &str, options: AddRowOptions) -> FormResult<String> {
        let (field, rows) = self.collection(path)?;
        if field
            .spec
            .get("multiple")
            .and_then(|v| v.get("max"))
            .and_then(Value::as_f64)
            .is_some_and(|max| rows.len() as f64 >= max)
        {
            return Err(FormError::input(format!(
                "Maximum row count reached: {path}"
            )));
        }
        let key = match options.key {
            Some(key) => key,
            None => fresh_key(rows)?,
        };
        check_key(&key)?;
        if rows.contains_key(&key) {
            return Err(FormError::input(format!("Row key already exists: {key}")));
        }
        let mut entries = rows
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect::<Vec<_>>();
        let at = match &options.after_key {
            None => entries.len(),
            Some(after) => {
                entries
                    .iter()
                    .position(|(k, _)| k == after)
                    .ok_or_else(|| FormError::input(format!("Unknown row: {after}")))?
                    + 1
            }
        };
        entries.insert(
            at,
            (key.clone(), normalize_row(field, options.value.as_ref())?),
        );
        self.commit(put_at(
            &self.data,
            &checked_segments(path)?,
            Value::Object(entries.into_iter().collect()),
        ))?;
        Ok(key)
    }

    /// Copy current values and generate new identities for nested repeated rows.
    pub fn copy_row(
        &mut self,
        path: &str,
        key: &str,
        mut options: AddRowOptions,
    ) -> FormResult<String> {
        let (field, rows) = self.collection(path)?;
        let value = rows
            .get(key)
            .ok_or_else(|| FormError::input(format!("Unknown row: {key}")))?;
        options.value = Some(copy_row_value(field, value)?);
        if options.after_key.is_none() {
            options.after_key = Some(key.into());
        }
        self.add_row(path, options)
    }

    /// Remove a row while enforcing the minimum count.
    pub fn remove_row(&mut self, path: &str, key: &str) -> FormResult<()> {
        let (field, rows) = self.collection(path)?;
        if !rows.contains_key(key) {
            return Err(FormError::input(format!("Unknown row: {key}")));
        }
        if field
            .spec
            .get("multiple")
            .and_then(|v| v.get("min"))
            .and_then(Value::as_f64)
            .is_some_and(|min| rows.len() as f64 <= min)
        {
            return Err(FormError::input(format!(
                "Minimum row count reached: {path}"
            )));
        }
        let mut rows = rows.clone();
        rows.shift_remove(key);
        self.commit(put_at(&self.data, &checked_segments(path)?, rows.into()))
    }

    /// Move a row without changing its identity or values.
    pub fn move_row(&mut self, path: &str, key: &str, index: usize) -> FormResult<()> {
        let (_, rows) = self.collection(path)?;
        let mut entries = rows
            .iter()
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect::<Vec<_>>();
        let from = entries
            .iter()
            .position(|(k, _)| k == key)
            .ok_or_else(|| FormError::input(format!("Unknown row: {key}")))?;
        if index >= entries.len() {
            return Err(FormError::input(format!("Invalid row position: {index}")));
        }
        if from == index {
            return Ok(());
        }
        let row = entries.remove(from);
        entries.insert(index, row);
        self.commit(put_at(
            &self.data,
            &checked_segments(path)?,
            Value::Object(entries.into_iter().collect()),
        ))
    }

    /// Apply a saved sequence key to one row and update its descendant paths.
    pub fn rekey_row(&mut self, path: &str, old_key: &str, new_key: &str) -> FormResult<()> {
        let (_, rows) = self.collection(path)?;
        check_key(new_key)?;
        if !rows.contains_key(old_key) {
            return Err(FormError::input(format!("Unknown row: {old_key}")));
        }
        if old_key == new_key {
            return Ok(());
        }
        if rows.contains_key(new_key) {
            return Err(FormError::input(format!(
                "Row key already exists: {new_key}"
            )));
        }
        let rows = rows
            .iter()
            .map(|(k, v)| {
                (
                    if k == old_key {
                        new_key.to_owned()
                    } else {
                        k.clone()
                    },
                    v.clone(),
                )
            })
            .collect::<Map<_, _>>();
        self.commit(put_at(&self.data, &checked_segments(path)?, rows.into()))
    }
}
